### PR TITLE
Add formatter to @dataform/api.

### PR DESCRIPTION
**Pull request by BenBirt**
for dataform-co/dataform-co#142

### CODE REVIEW

File: api/commands/format.ts [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+  return postProcessFormattedSqlx(
+    getIndividualSqlxStatements(node.contents)
+      .map(individualStatement => {
+        const placeholders: {
+          [placeholderId: string]: ISyntaxTreeNode;
+        } = {};
+        const unformattedPlaceholderSql = individualStatement
+          .map(child => {
+            if (typeof child !== "string") {
+              const placeholderId = crypto.randomBytes(16).toString("hex");
```
Comment (lewish) [inline api/commands/format.ts]: Can we seed this easily so it's consistent?

File: api/commands/format.ts [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+      ? wholeLineContainingPlaceholderId.slice(0, leftOfPlaceholder.length + placeholderId.length)
+      : placeholderId;
+  return sqlx.replace(textToReplace, formattedChild);
+}
+
+function formatChildSyntaxTreeNode(node: ISyntaxTreeNode, jsIndent: number): string {
+  if (node.startTokenType.endsWith("_startJsPlaceholder")) {
+    return formatJavaScriptPlaceholder(node, jsIndent);
+  }
+  if (node.startTokenType.endsWith("_startJs") || node.startTokenType.endsWith("_startConfig")) {
```
Comment (lewish) [inline api/commands/format.ts]: ditto, feels like these tokens would benefit from some more structure (e.g. contentType: "js" | "sql")

File: api/commands/format.ts [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        );
+      })
+      .join("\n\n---\n\n")
+  );
+}
+
+function getIndividualSqlxStatements(nodeContents: Array<string | ISyntaxTreeNode>) {
+  const sqlxStatements: Array<Array<string | ISyntaxTreeNode>> = [[]];
+  nodeContents.forEach(child => {
+    if (typeof child !== "string" && child.startTokenType.endsWith("_statementSeparator")) {
```
Comment (lewish) [inline api/commands/format.ts]: Can we reference the _statementSeperator from where it's defined

File: api/commands/format.ts [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+function formatChildSyntaxTreeNode(node: ISyntaxTreeNode, jsIndent: number): string {
+  if (node.startTokenType.endsWith("_startJsPlaceholder")) {
+    return formatJavaScriptPlaceholder(node, jsIndent);
+  }
+  if (node.startTokenType.endsWith("_startJs") || node.startTokenType.endsWith("_startConfig")) {
+    return `\n\n${formatJavaScript(concatenateSyntaxTreeContents(node))}\n\n`;
+  }
+  // This node must be an "inner" SQL block, e.g. "pre_operations { ... }",
```
Comment (lewish) [inline api/commands/format.ts]: Should we be explicit here? We will inevitably add new types that might not wanted to be treated like this. Might be better to just leave such blocks as they were

File: api/commands/format.ts (hunk 2 above)
Comment (BenBirt) [reply api/commands/format.ts]: yup, good call, done something very close to that.

File: api/commands/format.ts (hunk 3 above)
Comment (BenBirt) [reply api/commands/format.ts]: fixed as per your other comment

File: api/commands/format.ts (hunk 4 above)
Comment (BenBirt) [reply api/commands/format.ts]: done!

File: api/commands/format.ts (hunk 1 above)
Comment (BenBirt) [reply api/commands/format.ts]: I couldn't find a way to seed random numbers in node short of depending on another package :(

**Pull request merged**